lib/openstack_query/README.md

```diff
@@ -0,0 +1,375 @@
+```
+
+## Selecting external props using append_from()
+
+Find Errored and Shutoff VMs
+- AND haven't been updated in 60 days
+- AND the user who made the VM must belong to a specific user domain.
+- AND include the PROJECT_NAME
+
+```python
+from openstack_query import UserQuery
+from enums.query.props.server_properties import ServerProperties
+from enums.query.props.user_properties import UserProperties
+from enums.query.props.project_properties import ProjectProperties
+from enums.query.query_presets import QueryPresetsGeneric, QueryPresetsDateTime
+
+
+# setup and run a User Query to get user info for all users in a specific domain
+user_query = UserQuery()
+user_query.select(UserProperties.USER_NAME, UserProperties.USER_EMAIL)
+user_query.where(QueryPresetsGeneric.EQUAL_TO, UserProperties.USER_DOMAIN_ID, value="user-domain-id1")
+user_query.run("prod")
+
+server_query = user_query.then("SERVER_QUERY", keep_previous_results=True)
+server_query.select(ServerProperties.SERVER_NAME, ServerProperties.SERVER_ID)
+server_query.where(QueryPresetsGeneric.ANY_IN, ServerProperties.SERVER_STATUS, values=["ERROR", "SHUTOFF"])
+server_query.where(QueryPresetsDateTime.OLDER_THAN, ServerProperties.SERVER_LAST_UPDATED_DATE, days=60)
+server_query.run("prod", as_admin=True, all_projects=True)
+
+# We need to get project name by running append_from()
+# append_from() command below is the same as doing the following:
+#   project_ids = server_query.group_by(ServerProperties.PROJECT_ID).to_list().keys()
+#   p = ProjectQuery().select(ProjectProperties.PROJECT_NAME).where(
+#        QueryPresetsGeneric.ANY_IN, ProjectProperties.PROJECT_ID, values=list(project_ids)
+#   )
+#   p.run("prod")
+#   res = p.to_list()
+# `res` is then combined zipped together into the current output
+server_query.append_from("PROJECT_QUERY", "prod", ProjectProperties.PROJECT_ID)
+
+# Note it's not possible to group by external properties (yet)
+server_query.group_by(ServerProperties.PROJECT_ID)
+
+# This will print out results and includes project name for every VM too
+server_query.to_string()
+```
+
+### NOTE - requirement to use enums for properties and query presets will be removed soon - see issue [88](https://github.com/stfc/st2-cloud-pack/issues/88)
+
+
+# Outputting
+
+you can output the results of your query in the following ways:
+
+`to_list()` - output the info to a list or a dict (if grouped)
+
+**optional params:**
+
+- `as_objects`: `bool`
+    - if True will return results as Openstack Objects instead of selected for properties
+      - if group_by() was run - then the results will be a dictionary
+    - if False will return selected for results as normal
+
+
+- `flatten`: `bool`
+
+If True will flatten the results by selected for properties:
+
+Instead of returning:
+```python
+print(query.to_list(flatten=False))
+
+[
+
+  # first result
+  { "project_name": "foo", "project_id": "bar" },
+
+  # second result
+  { "project_name": "foo1", "project_id": "bar2" },
```

Comment:
Second result: "bar2" looks like it should be "bar1" to match "foo1".

Reply:
heh yea - changing